pygls/workspace.py

```diff
@@ -54,15 +54,26 @@ def __str__(self):
         return str(self.uri)
 
     def _apply_incremental_change(self, change: TextDocumentContentChangeEvent) -> None:
-        """Apply an incremental text change to the document."""
+        """Apply an either INCREMENTAL or FULL text change to the document
+
+        Note on INCREMENTAL versus FULL:
+            Even if a server accepts INCREMENTAL SyncKinds, clients may request
+            a FULL SyncKind. In LSP 3.x, clients make this request by omitting
+            both Range and RangeLength from their request. Consequently, the
+            attributes "range" and "range_length" will be missing from FULL
+            content update client requests in the pygls Python library.
+        """
+        if (
```

Comment:
I didn't know that, but if that's the case, I think we should keep logic on which text document synchronization method to keep in `apply_change` method.

```python
def apply_change(self, change: TextDocumentContentChangeEvent) -> None:
    """Apply change to a document, depending on synchronization kind."""
    client_sync_kind = TextDocumentSyncKind.FULL if not (hasattr(change, 'range') or hasattr(change, 'range_length')) else TextDocumentSyncKind.INCREMENTAL
    if self._sync_func is None:
        self._sync_func = {
            TextDocumentSyncKind.FULL: self._apply_full_change,
            TextDocumentSyncKind.INCREMENTAL: self._apply_incremental_change
        }.get(min(client_sync_kind, self._server_sync_kind), None)
        
    try:
        self._sync_func(change)
    except TypeError:
        # Don't sync document if sync kind is TextDocumentSyncKind.NONE
        pass 
```

And you will have to slightly modify `__init__` method. What do you think?

Reply:
Great suggestion, I'll work on this tomorrow afternoon and send an update then!

Reply:
@danixeee I've made your suggested updates (with some minor changes). I've put lots of information in the docstrings / comments / commit message, but please let me know if you have any questions!